pkg/watcher/rated.go

```diff
@@ -25,83 +28,122 @@ type getValidatorEffectivenessData struct {
 
-	body, err := io.ReadAll(res.Body)
-	if err != nil {
-		log.WithError(err).WithFields(log.Fields{
-			"url":            url,
-			"validation-key": key,
-		}).Warn("unable to read rated network http body")
+			// Abort on unknown error
+			return nil, err
+		}
 
-		return nil, err
-	}
+		defer res.Body.Close()
 
-	log.WithFields(log.Fields{
-		"url":            url,
-		"validation-key": key,
-	}).Info("parsing response from rated network")
-
-	var response getValidationEffectiveness
-	err = json.Unmarshal(body, &response)
-	if err != nil {
-		log.WithError(err).WithFields(log.Fields{
-			"url":            url,
-			"validation-key": key,
-		}).Warn("unable to parse rated network http body into expected response")
-
-		return nil, err
-	}
-
-	if len(response.Data) != 1 {
 		log.WithFields(log.Fields{
 			"url":            url,
-			"status-code":    res.StatusCode,
 			"validation-key": key,
-			"nb-entries":     len(response.Data),
-		}).Warn("expected 1 entry of statistics for the validation key")
+		}).Info("Reading response from rated network")
+
+		if res.StatusCode == 500 {
+			log.WithFields(log.Fields{
+				"url":           url,
+				"status-code":   res.StatusCode,
+				"validator-key": key,
+			}).Error("Unable to fetch validator data due to internal server error. Aborting")
+
+			return nil, fmt.Errorf("Internal server error received from rated network")
+		}
+
+		if res.StatusCode == 404 {
+			log.WithFields(log.Fields{
+				"url":            url,
+				"status-code":    res.StatusCode,
+				"validation-key": key,
+			}).Warn("Validator not found")
+
+			return nil, fmt.Errorf("Validator not found")
+		}
+
+		if res.StatusCode == 200 {
+			body, err := io.ReadAll(res.Body)
+			if err != nil {
+				log.WithError(err).WithFields(log.Fields{
+					"url":            url,
+					"validation-key": key,
+				}).Warn("Unable to read rated network http body")
+
+				return nil, err
+			}
+
+			log.WithFields(log.Fields{
+				"url":            url,
+				"validation-key": key,
+			}).Info("Parsing response from rated network")
+
+			var response getValidationEffectiveness
+			err = json.Unmarshal(body, &response)
+			if err != nil {
+				log.WithError(err).WithFields(log.Fields{
+					"url":            url,
+					"validation-key": key,
+				}).Warn("Unable to parse rated network http body into expected response")
+
+				return nil, err
+			}
+
+			if len(response.Data) != 1 {
+				log.WithFields(log.Fields{
+					"url":            url,
+					"status-code":    res.StatusCode,
+					"validation-key": key,
+					"nb-entries":     len(response.Data),
+				}).Warn("Expected 1 entry of statistics for the validation key")
+
+				return nil, fmt.Errorf("unexpected data response from rated network")
+			} else {
+				log.WithFields(log.Fields{
+					"url":            url,
+					"validation-key": key,
+				}).Info("Successfully fetched statistics for validator")
+
+				return &response.Data[0], nil
+
+			}
+
+		}
+
+		if res.StatusCode == 429 {
```

Comment:
Maybe let's exponential backoff by default? if we receive here a 401, or something we didn't expect, we'll query 10 times the API as fast as we can.

Reply:
There is no other possible response code atm, but yeah you are right it needs handling.
Actually, it needs to fail fast in any unknown status code. 

I will add another commit. Thanks!